### PR TITLE
Update to openapi-core 0.17

### PR DESCRIPTION
**Pull request by zupo**
This is quite a breaking change. 

We are dropping support for all older versions of openapi-core and for OpenAPI 3.0 (upgraded to 3.1).

Refs #203
Refs https://github.com/python-openapi/openapi-core/issues/542

### CODE REVIEW

File: py38/pyproject.toml [hunk 1]
```diff
@@ -26,7 +26,8 @@ exclude = ["pyramid_openapi3/tests/"]
 [tool.poetry.dependencies]
 python = "^3.8"
 
-openapi-core = "==0.16.1"
+# openapi-core = "==0.16.1"
+openapi-core = { git = "https://github.com/zupo/openapi-core.git", branch = "patch-2" }
```
Comment (zupo) [inline py38/pyproject.toml:30]: See: https://github.com/python-openapi/openapi-core/pull/544

File: pyproject.toml [hunk 2]
```diff
@@ -26,7 +26,8 @@ exclude = ["pyramid_openapi3/tests/"]
 [tool.poetry.dependencies]
 python = "^3.8"
 
-openapi-core = ">=0.16.1"
+# openapi-core = ">=0.16.1"
+openapi-core = { git = "https://github.com/zupo/openapi-core.git", branch = "patch-2" }
```
Comment (zupo) [inline pyproject.toml:30]: See https://github.com/python-openapi/openapi-core/pull/544.

File: pyramid_openapi3/__init__.py [hunk 3]
```diff
@@ -86,9 +80,7 @@ def openapi_validated(request: Request) -> dict:
 
     if request.environ.get("pyramid_openapi3.validate_request"):
         openapi_request = PyramidOpenAPIRequest(request)
-        validated = settings["request_validator"].validate(
-            settings["spec"], openapi_request
-        )
+        validated = settings["request_validator"].unmarshal(openapi_request)
```
Comment (zupo) [inline pyramid_openapi3/__init__.py:83]: The `spec` is now provided when `settings["request_validator"]` is instantiated during Pyramid bootup.

File: pyramid_openapi3/tests/test_app_construction.py [hunk 4]
```diff
@@ -39,17 +39,30 @@
...
+    openapi: "3.1.0"
     info:
       version: "1.0.0"
       title: Foo API
     servers:
       - url: /api/v1
     paths:
       /foo:
-        $ref: "paths.yaml#/foo"
+        # $ref: "paths.yaml#/foo"
```
Comment (zupo) [inline pyramid_openapi3/tests/test_app_construction.py:50]: Can't get `refs` to work. Needs further investigation.

File: pyramid_openapi3/tests/test_extract_errors.py [hunk 5]
```diff
@@ -159,9 +161,8 @@ def test_invalid_path_parameter_regex(self) -> None:
         assert res.json == [
             {
                 "exception": "ValidationError",
+                "field": "bar",
                 "message": "'not-a-valid-uuid' does not match '^[0-9]{2}-[A-F]{4}$'",
-                # TODO: ideally, this response would include "field"
-                # but I don't know how I can achieve this ATM ¯\_(ツ)_/¯  # noqa: ENC100
```
Comment (zupo) [inline pyramid_openapi3/tests/test_extract_errors.py:164]: 💪

File: pyramid_openapi3/tests/test_extract_errors.py [hunk 6]
```diff
@@ -565,30 +566,29 @@ def hello(self, context: t.Any, request: Request) -> str:
...
-                            type="unique-name",
-                            original_exception=Exception(
-                                f"Name '{name}' already taken. Choose a different name!"
-                            ),
-                            field="name",
-                        )
-                    ]
-                )
-            return True
+    def validate(self, name: str) -> bool:
```
Comment (zupo) [inline pyramid_openapi3/tests/test_extract_errors.py:569]: Same thing as before, just no longer needs to be wrapped in a class.

File: pyramid_openapi3/tween.py [hunk 7]
```diff
@@ -41,8 +41,8 @@ def excview_tween(request: Request) -> Response:
             if "routes" in gsettings:
                 settings_key = gsettings["routes"][request.matched_route.name]
                 settings = request.registry.settings[settings_key]
-            result = settings["response_validator"].validate(
-                settings["spec"], request=openapi_request, response=openapi_response
+            result = settings["response_validator"].unmarshal(
```
Comment (zupo) [inline pyramid_openapi3/tween.py:44]: We now have to use `unmarshal` to get the deserialized values, `validate` doesn't return anything.

In this light the `request.openapi_validated` should maybe be renamed to `request.openapi_unmarshalled` but that's such a mouthful.

**Comment by zupo** [comment]
@am-on, @damonhook, @dz0ny, @miohtama: if you can spare some brain cycles, I'd love a review of the changes.

File: pyramid_openapi3/tests/test_app_construction.py (hunk 4 above)
Comment (zupo) [reply pyramid_openapi3/tests/test_app_construction.py:50]: So it seems `refs` work for everything but for paths? 

Details: https://github.com/Pylons/pyramid_openapi3/pull/206/commits/9b080c1282358f290c04e0234f5ee6e26bd86678

Comment (zupo) [reply pyramid_openapi3/tests/test_app_construction.py:50]: Is this good enough?

Comment (zupo) [reply pyramid_openapi3/tests/test_app_construction.py:50]: @damonhook: can you pls have a look if you can get `examples/splitfile` to work with paths being in a separate `paths.yaml` file? I can probably take it from there then, fix the rests of the tests and such.

File: pyramid_openapi3/__init__.py [hunk 8]
```diff
@@ -212,7 +203,7 @@ def add_spec_view(
     permission: str = NO_PERMISSION_REQUIRED,
     apiname: str = "pyramid_openapi3",
 ) -> None:
-    """Serve and register OpenApi 3.0 specification file.
+    """Serve and register OpenApi 3.1 specification file.
```
[the remark this answers is not in the document]
Comment (zupo) [reply pyramid_openapi3/__init__.py:206]: Ugh. I guess I have to revert all my changes back to OpenAPi 3.0.

[the remark this answers is not in the document]
Comment (zupo) [reply pyramid_openapi3/__init__.py:206]: I've bumped the Swagger UI version in this PR to `5.0.0-alpha.6`.

File: pyproject.toml (hunk 2 above)
Comment (zupo) [reply pyproject.toml]: My fixed was merged, I updated this PR back to using upstream. 

I'll pin to a specific version once a new release of openapi-core is cut.

**Review by damonhook — COMMENTED** [review]
I noticed that with a later release of `openapi-core`, running the `Spec.create(...)` seems to run the validator https://github.com/python-openapi/openapi-core/blob/59996250c5f316c8fc64512115bfd18f587b78ee/openapi_core/spec/paths.py#L29

So we can likely remove the `validate_spec()` calls we make 
[here](https://github.com/Pylons/pyramid_openapi3/blob/389356bbed0b583eb2e6ed032dca42ac39fbc06f/pyramid_openapi3/__init__.py#L226) and [here](https://github.com/Pylons/pyramid_openapi3/blob/389356bbed0b583eb2e6ed032dca42ac39fbc06f/pyramid_openapi3/__init__.py#L278). Right now we are validating the spec twice

File: pyramid_openapi3/tests/test_app_construction.py (hunk 4 above)
Comment (damonhook) [reply pyramid_openapi3/tests/test_app_construction.py:50]: https://github.com/python-openapi/openapi-spec-validator/issues/207

**Comment by zupo** [comment]
Swagger UI 5 is released! https://github.com/swagger-api/swagger-ui/releases

File: pyramid_openapi3/tests/test_app_construction.py (hunk 4 above)
Comment (damonhook) [reply pyramid_openapi3/tests/test_app_construction.py:50]: I see that this `openapi-spec-validator` issue has been closed. The refs in the paths should hopefully work now

<details><summary>Example openapi.yaml</summary>
<p>

```yaml
openapi: 3.1.0

info:
  title: Example API
  version: 1.0.0

paths:
  /:
    $ref: "#/components/pathItems/rootPath"

components:
  pathItems:
    rootPath:
      get:
        responses:
          200:
            description: Example Response⏎
```

</p>
</details> 

```bash
openapi-spec-validator openapi.yaml

OK
```
Previous version of `openapi-spec-validator` was raising an error

Will give it a test with the `examples/splitfile`

**Comment by kskarthik** [comment]
> Swagger UI 5 is released! https://github.com/swagger-api/swagger-ui/releases

do hey have multiple supported versions? I have observed 4.x series & 5.x series

**Review by damonhook — COMMENTED** [review]
I see from the 0.17 release notes that `Spec.create` is deprecated.
https://github.com/python-openapi/openapi-core/releases/tag/0.17.0

While it will still work on 0.17, It does have a deprecation warning _(and it has been removed in 0.18)_

### In `add_spec_view`
https://github.com/Pylons/pyramid_openapi3/blob/389356bbed0b583eb2e6ed032dca42ac39fbc06f/pyramid_openapi3/__init__.py#L227

Should change to
```python 
spec = Spec.from_dict(spec_dict)
```

### In `add_spec_view_directory`
https://github.com/Pylons/pyramid_openapi3/blob/389356bbed0b583eb2e6ed032dca42ac39fbc06f/pyramid_openapi3/__init__.py#L279

Should change to

```python
spec = Spec.from_dict(spec_dict, spec_url=spec_url)
```

> [!NOTE]
> In 0.18, `spec_url` will be changing to `base_uri`, but for 0.17 we need to still use `spec_url`

**Comment by zupo** [comment]
The upgrade to openapi-core was done in https://github.com/Pylons/pyramid_openapi3/pull/220. 

This PR then remains a good basis to work on bringing in [OpenAPI v3.1 support](https://github.com/Pylons/pyramid_openapi3/issues/196).

**Pull request closed without merging**